File: web/lib/types/goals.ts
// Mirrors backend DTOs from backend/api/.../api/goals/dto/*.java
// Keep in sync when the backend record shapes change.

export type GoalDomain =
  | "CARDIOVASCULAR"
  | "BODY_COMPOSITION"
  | "STRENGTH"
  | "METABOLIC"
  | "SLEEP"
  | "LONGEVITY"
  | "OTHER";

export type GoalStatus = "ACTIVE" | "COMPLETED" | "ARCHIVED";

export type GoalSource = "MANUAL" | "AI_GENERATED" | "AI_ASSISTED";

export type PhaseStatus = "LOCKED" | "ACTIVE" | "COMPLETED";

export type StepKind = "MANUAL" | "THRESHOLD" | "SUSTAINED" | "COUNT";

export type Comparator = "LT" | "LTE" | "GT" | "GTE" | "EQ";

export type StepMetricBinding = {
  metricKey: string;
  comparator: Comparator;
  targetValue: number;
  windowDays: number | null;
  countFrom: string | null;
};

export type GoalResponse = {
  goalId: string;
  title: string;
  description: string;
  domain: GoalDomain;
  status: GoalStatus;
  startDate: string;
  targetDate: string;
  createdAt: string;
  updatedAt: string;
  completedAt: string | null;
  phaseOrder: string[];
  source: GoalSource;
};

export type StepResponse = {
  stepId: string;
  phaseId: string;
  goalId: string;
  title: string;
  orderIndex: number;
  kind: StepKind;
  done: boolean;
  doneAt: string | null;
  manualOverride: boolean;
  metric: StepMetricBinding | null;
  metricRegressed: boolean | null;
};

export type PhaseResponse = {
  phaseId: string;
  goalId: string;
  title: string;
  description: string;
  orderIndex: number;
  status: PhaseStatus;
  targetStartDate: string;
  targetEndDate: string;
  completedAt: string | null;
  stepOrder: string[];
  steps: StepResponse[];
};

export type GoalDeepResponse = GoalResponse & {
  phases: PhaseResponse[];
};

export const DOMAIN_LABEL: Record<GoalDomain, string> = {
  CARDIOVASCULAR: "Cardiovascular",
  BODY_COMPOSITION: "Body composition",
  STRENGTH: "Strength",
  METABOLIC: "Metabolic",
  SLEEP: "Sleep",
  LONGEVITY: "Longevity",
  OTHER: "Other",
};

export const DOMAINS: GoalDomain[] = [
  "CARDIOVASCULAR",
  "BODY_COMPOSITION",
  "STRENGTH",
  "METABOLIC",
  "SLEEP",
  "LONGEVITY",
  "OTHER",
];

export const STEP_KINDS: StepKind[] = [
  "MANUAL",
  "THRESHOLD",
  "SUSTAINED",
  "COUNT",
];

export const STEP_KIND_LABEL: Record<StepKind, string> = {
  MANUAL: "Manual",
  THRESHOLD: "Threshold",
  SUSTAINED: "Sustained",
  COUNT: "Count",
};

export const COMPARATOR_SYMBOL: Record<Comparator, string> = {
  LT: "<",
  LTE: "≤",
  GT: ">",
  GTE: "≥",
  EQ: "=",
};

// The fixed metric-key registry — the contract between Goals and the
// other modules (IMPL-12 spec "Metric key registry"). `lowerIsBetter`
// drives the RangeIndicator good/alert coloring for threshold readouts;
// `unit` is display-only.
export type MetricKeyMeta = {
  key: string;
  label: string;
  unit: string;
  lowerIsBetter: boolean;
};

export const METRIC_REGISTRY: MetricKeyMeta[] = [
  { key: "body.weight", label: "Body weight", unit: "lb", lowerIsBetter: true },
  { key: "body.bodyFatPct", label: "Body fat %", unit: "%", lowerIsBetter: true },
  { key: "body.leanMass", label: "Lean mass", unit: "lb", lowerIsBetter: false },
  { key: "blood.ldl", label: "LDL", unit: "mg/dL", lowerIsBetter: true },
  { key: "blood.apoB", label: "ApoB", unit: "mg/dL", lowerIsBetter: true },
  { key: "blood.hba1c", label: "HbA1c", unit: "%", lowerIsBetter: true },
  { key: "blood.hsCRP", label: "hs-CRP", unit: "mg/L", lowerIsBetter: true },
  { key: "vitals.restingHr", label: "Resting HR", unit: "bpm", lowerIsBetter: true },
  { key: "vitals.hrv", label: "HRV", unit: "ms", lowerIsBetter: false },
  { key: "vitals.sleepScore", label: "Sleep score", unit: "", lowerIsBetter: false },
  { key: "workouts.count", label: "Workouts logged", unit: "", lowerIsBetter: false },
  { key: "workouts.weeklyVolume", label: "Weekly volume", unit: "lb", lowerIsBetter: false },
  { key: "nutrition.proteinAvg7d", label: "Protein (7d avg)", unit: "g", lowerIsBetter: false },
  { key: "nutrition.carbsAvg7d", label: "Carbs (7-day avg)", unit: "g", lowerIsBetter: false },
  { key: "nutrition.fatAvg7d", label: "Fat (7-day avg)", unit: "g", lowerIsBetter: false },
  { key: "nutrition.caloriesAvg7d", label: "Calories (7-day avg)", unit: "kcal", lowerIsBetter: false },
  { key: "meds.adherence30d", label: "Med adherence (30d)", unit: "%", lowerIsBetter: false },
];

export const METRIC_LABEL: Record<string, string> = Object.fromEntries(
  METRIC_REGISTRY.map((m) => [m.key, m.label]),
);

export function metricMeta(key: string): MetricKeyMeta | undefined {
  return METRIC_REGISTRY.find((m) => m.key === key);
}

File: android/core-domain/src/main/java/com/gte619n/healthfitness/domain/goals/Goals.kt
package com.gte619n.healthfitness.domain.goals

import com.squareup.moshi.Json

// Goals domain models (IMPL-12). These mirror the backend JSON contract
// exactly — see web/lib/types/goals.ts (GoalResponse, GoalDeepResponse,
// PhaseResponse, StepResponse, StepMetricBinding) — and double as the Moshi
// wire types. Dates/timestamps are kept as ISO-8601 strings and parsed in the
// UI layer, per IMPL-12 remaining-assumptions item 18.

enum class GoalDomain {
    CARDIOVASCULAR,
    BODY_COMPOSITION,
    STRENGTH,
    METABOLIC,
    SLEEP,
    LONGEVITY,
    OTHER,
    ;

    val label: String
        get() = when (this) {
            CARDIOVASCULAR -> "Cardiovascular"
            BODY_COMPOSITION -> "Body composition"
            STRENGTH -> "Strength"
            METABOLIC -> "Metabolic"
            SLEEP -> "Sleep"
            LONGEVITY -> "Longevity"
            OTHER -> "Other"
        }
}

enum class GoalStatus { ACTIVE, COMPLETED, ARCHIVED }

enum class GoalSource { MANUAL, AI_GENERATED, AI_ASSISTED }

enum class PhaseStatus { LOCKED, ACTIVE, COMPLETED }

enum class StepKind { MANUAL, THRESHOLD, SUSTAINED, COUNT }

enum class Comparator {
    LT, LTE, GT, GTE, EQ;

    val symbol: String
        get() = when (this) {
            LT -> "<"
            LTE -> "≤"
            GT -> ">"
            GTE -> "≥"
            EQ -> "="
        }
}

data class StepMetricBinding(
    val metricKey: String,
    val comparator: Comparator,
    val targetValue: Double,
    val windowDays: Int? = null,
    val countFrom: String? = null,
)

data class Step(
    val stepId: String,
    val phaseId: String,
    val goalId: String,
    val title: String,
    val orderIndex: Int,
    val kind: StepKind,
    val done: Boolean,
    val doneAt: String? = null,
    val manualOverride: Boolean = false,
    val metric: StepMetricBinding? = null,
    val metricRegressed: Boolean? = null,
)

data class Phase(
    val phaseId: String,
    val goalId: String,
    val title: String,
    // Backend-optional fields kept nullable so a sparse phase can't fail the
    // deep-goal Moshi parse. The UI reads dates through parseDate()/
    // formatDateRange(), both String?-safe.
    val description: String? = null,
    val orderIndex: Int,
    val status: PhaseStatus,
    val targetStartDate: String? = null,
    val targetEndDate: String? = null,
    val completedAt: String? = null,
    val stepOrder: List<String> = emptyList(),
    val steps: List<Step> = emptyList(),
)

/** Shallow goal — the list/card shape (no phases). */
data class Goal(
    val goalId: String,
    val title: String,
    // Optional on the backend (CreateGoalRequest doesn't require it) and the
    // PATCH response can echo it back null — keep nullable so a single sparse
    // goal doesn't fail the whole `List<Goal>` Moshi parse and blank the list.
    val description: String? = null,
    val domain: GoalDomain,
    val status: GoalStatus,
    val startDate: String? = null,
    val targetDate: String? = null,
    // Server timestamps: the PATCH path returns `updatedAt` null (it's filled
    // by persistence on save, not echoed), and a just-created goal may read
    // back before its serverTimestamp() sentinel resolves. Nullable so Moshi
    // tolerates it; the UI never displays these.
    val createdAt: String? = null,
    val updatedAt: String? = null,
    val completedAt: String? = null,
    val phaseOrder: List<String> = emptyList(),
    val source: GoalSource,
)

/** Deep goal — the roadmap shape (phases + steps). Mirrors GoalDeepResponse. */
data class GoalDeep(
    val goalId: String,
    val title: String,
    val description: String? = null,
    val domain: GoalDomain,
    val status: GoalStatus,
    val startDate: String? = null,
    val targetDate: String? = null,
    val createdAt: String? = null,
    val updatedAt: String? = null,
    val completedAt: String? = null,
    val phaseOrder: List<String> = emptyList(),
    val source: GoalSource,
    val phases: List<Phase> = emptyList(),
)

/** Body for PATCH .../steps/{sid}. Either toggle done, or reset to auto. */
data class StepPatchRequest(
    @Json(name = "done") val done: Boolean? = null,
    @Json(name = "resetToAuto") val resetToAuto: Boolean? = null,
)
